homeassistant/components/hydrawise/coordinator.py

```diff
@@ -2,27 +2,46 @@
 
 from __future__ import annotations
 
+from dataclasses import dataclass
 from datetime import timedelta
 
 from pydrawise import HydrawiseBase
-from pydrawise.schema import User
+from pydrawise.schema import Controller, User, Zone
 
 from homeassistant.core import HomeAssistant
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
 
 from .const import DOMAIN, LOGGER
 
 
-class HydrawiseDataUpdateCoordinator(DataUpdateCoordinator[User]):
+@dataclass
+class HydrawiseData:
+    """Container for data fetched from the Hydrawise API."""
+
+    user: User
+    controllers: dict[int, Controller]
+    zones: dict[int, Zone]
+
+
+class HydrawiseDataUpdateCoordinator(DataUpdateCoordinator[HydrawiseData]):
     """The Hydrawise Data Update Coordinator."""
 
+    api: HydrawiseBase
+
     def __init__(
         self, hass: HomeAssistant, api: HydrawiseBase, scan_interval: timedelta
     ) -> None:
         """Initialize HydrawiseDataUpdateCoordinator."""
         super().__init__(hass, LOGGER, name=DOMAIN, update_interval=scan_interval)
         self.api = api
 
-    async def _async_update_data(self) -> User:
+    async def _async_update_data(self) -> HydrawiseData:
         """Fetch the latest data from Hydrawise."""
-        return await self.api.get_user()
+        user = await self.api.get_user()
+        controllers = {}
+        zones = {}
+        for controller in user.controllers:
```

Comment:
As you are the maintainer of the library, can we do it there?

Something like 
```python
class User:
    ...
    controller: dict[int, Controller]
    ...

class Controller:
    ...
    zones: dict[int, Zone]
    ...
```

Reply:
I'm not sure I agree with this. The purpose of the `HydrawiseData` container is to map the API data in a way that is convenient for HA to consume. For example, I have an upcoming PR that will make additional calls to the API to compute water use by zone. This data is computed and there is no 1-to-1 mapping from the API itself.

The mapping from API data to what is convenient for HA to consume is business logic that does not really belong in the API library. Otherwise, if we have 10 other clients consuming this library, we will end up with 10 different kind of fields for 10 different clients.

Reply:
Currently, it looks like the library is used only by HA, but the author can give us more insights there.
Looking at the integration code, you only use one zone or controller at a time.
So maybe also the other clients are looping over all zones to find the correct one too. If this will happen then moving the change to the library is reasonable.
If you a client wants still the list, he can always call `.values()`

Reply:
If we moved this logic to the library, then how would you deal with the following (coming in the next PR):
```
        daily_water_use: dict[int, DailyWaterUse] = {}
        for controller in user.controllers:
            controllers[controller.id] = controller
            for zone in controller.zones:
                zones[zone.id] = zone
            for sensor in controller.sensors:
                sensors[sensor.id] = sensor
                if "flow meter" in sensor.model.name.lower():
                    await self._async_update_daily_water_use(
                        controller, sensor, daily_water_use
                    )
```
where `_async_update_daily_water_use` is making two additional calls to the API? Is your preference for all this logic to be in the library? And then have the `User` object to have methods `get_controller_by_id()`, `get_zone_by_id()`, and `get_sensor_by_id()`?

Reply:
I'd rather not move this into the library. While the primary reason I wrote the library was for HA, I don't know if there are other users and changing the field type would be a breaking change.

Reply:
Okay, the change is fine.


One quick notive for the future PRs:
We don't allow device and service integrations to calculate state, and `_async_update_daily_water_use` could do some aggregation in the integration, which is not allowed.
Also, API calls should be made only for enabled entities. For example, I have disabled the flow meter sensor entity in HA, then the coordinator should not make the additional API call.